doc/zk-integration.md

```diff
@@ -0,0 +1,214 @@
+                commTxsRollup,
+            ];
+
+            SnarkVerify(
+                publicInputs,
+                snarkProof,
+                vkRollup); // This verification key corresponds to the circuit depicted in Figure 1.
+        );
+
+        } else { // No escape hatch, use the state of Espresso consensus
+
+            (newEspressoState,  blockNumberEspresso) = lcContract.getLastBlockCommitment();
+
+
+            if (blockNumberEspresso <= lastBlockNumberEspresso) {
+                revert();
+            }
+
+            bytes[] publicInputs = [
+                previousVMState,
+                newVMState,
+                commTxsRollup,
+                previousEspressoState,
+                newEspressoState
+            ];
+
+            SnarkVerify(
+                publicInputs,
+                snarkProof,
+                vkEspresso // This verification key corresponds to the circuit depicted in Figure 2.
+            );
+
+            previousEspressoState = newEspressoState;
+            lastBlockNumberEspresso =  blockNumberEspresso;
+
+        }
+
+        previousVMState = newVMState;
+
+    }
+}
+
+```
+
+## Integration 2: Verify Espresso consensus inside the rollup circuit
+
+![image](zk-rollup-circuit.svg) **Figure 3:** Rollup circuit with additional gadgets _Espresso Consensus_, _Collect &
+Filter_, and _COMMs Equivalence_. Private inputs of the circuit are written in bold font.
+
+The circuit depicted in Figure 3 operates as follows:
+
+- The _Espresso Consensus_ gadget checks that the block commitment for Espresso block _BLOCK_NUMBER_ESP_ is
+  _BLOCK_COMM_ESP_, using the multi-signature _STATE_SIGS_ESP_.
+- The _Collect & Filter_ gadget receives as input _BLOCK_COMM_ESP_NEW_ which is the commitment to the latest Espresso
+  block available and _BLOCK_COMM_ESP_OLD_. Both of these commitments are public inputs. The first witness of this
+  circuit is _COMM_TXS_HISTORY_ESP_ which is a commitment to all the rollup transactions that have been sequenced since
+  the last Espresso state update _BLOCK_COMM_ESP_OLD_. The relationship between _BLOCK_COMM_ESP_, _BLOCK_COMM_ESP_OLD_,
+  and _COMM_TXS_HISTORY_ESP_ can be checked using a second witness _PROOF_TXS_HISTORY_ESP_.
+- The _COMMs Equivalence_ gadget checks that using the same rollup inputs _ROLLUP_TXS_, we obtain _COMM_TXS_HISTORY_ESP_
+  using the Espresso commitment scheme for representing a set of transactions and the commitment _COMM_TXS_ROLLUPS_ that
+  is used the by _zkVM_ gadget.
+- The zkVM gadget is the original gadget of the rollup circuit that proves a correct transition from state
+  _COMM_STATE_VM i_ to the next state _COMM_STATE_VM i+1_ when applying the transactions represented by the commitment
+  value _COMM_TXS_ROLLUP_.
+- These three gadgets above return a boolean: true if the verification succeeds and false otherwise.
+- For the circuit to accept, all these gadget outputs must be true, and thus we add an _AND_ gate.
+
+```solidity
```

Comment:
- As above, please explain the purpose of this code with a caption.
- I'm confused: why does snark verification happen inside a smart contract?

Reply:
> I'm confused: why does snark verification happen inside a smart contract?

Because it is the rollup contract that updates the rollup state in a provable way. Where would you expect this verification to happen instead?

Reply:
Code redundancy removed in e246dc999fc8c17fa98f25db7026e1945f1a3aa2.